### ISSUE TITLE
Automatically invoking On Cluster on local tables when DDL is executed on a Distributed table

### ISSUE THREAD

**Issue by schien-dong** [ClickHouse/ClickHouse#28912]
Setup:

Shard 1:
create table test_slice(x UInt32) Engine=ReplicatedMergeTree('/s1-r1','rep1') order by x;
create table test as test_slice engine=Distributed(mycluster, testdb, test_slice,rand());

Shard 2:

create table test_slice(x UInt32) Engine=ReplicatedMergeTree('/s1-r1','rep2') order by x;
create table test as test_slice engine=Distributed(mycluster, testdb, test_slice,rand());

Currently the Distributed table "test" neatly abstracts away the local tables. Insert and Select statements can run directly against the combined "test" table.

However, when deleting a record from the test table, "Mutation is not allowed" on a Distributed table. A distributed DDL is needed in this case:

alter table test_slice on cluster mycluster delete where 1;

This exposes the "local" tables and creates inconsistency to the application layer.

When an application is prototyped on a MergeTree table, the Updates and Deletes will have to be altered when moving to a cluster.

Suggested solution:
Allow a setting in clickhouse-server/config.xml, something like:
<cluster_ddl_forwarding>true</cluster_ddl_forwarding>

When this setting is enabled, this query:
    update table test delete where 1;
is interpreted as:
    update table test_slice on cluster mycluster delete where 1;

The Distributed table "test" has all the information to perform this rewrite.

**Comment by filimonov** [comment]
Did you check DatabaseReplicated?

https://github.com/ClickHouse/clickhouse-presentations/blob/master/meetup51/DatabaseReplicated.pdf
https://youtu.be/HPJOgzQkRls?t=5024

https://github.com/ClickHouse/ClickHouse/pull/16193

**Comment by schien-dong** [comment]
I read through the slide - is there a particular slide I'm missing? The feature request is for a syntactical sugar / convenient abstraction. The slides explained clustering and replication arrangement in detail. I don't think they are the same thing?

**Comment by filimonov** [comment]
> I read through the slide - is there a particular slide I'm missing? I don't think they are the same thing?

You're right. I've just misunderstood the feature you were asking for ("diagonal reading")

> The feature request is for a syntactical sugar / convenient abstraction. The slides explained clustering and replication arrangement in detail. 

Reread that. Sounds ok.

**Comment by schien-dong** [comment]
There is another edge case to handle: if the Distributed table is pointing to a Buffer table, the "redirect" should be further resolved to an eventual Replicated table. This is because mutation is not allowed on Buffer tables either. Alternatively, a Buffer table can independently "forward" the distributed DDL.

**Comment by tavplubix** [comment]
And how should it work if DIstributed table is pointing to, for example, View or Merge table?

**Comment by schien-dong** [comment]
There could be two approaches:

Option 1: Distributed will traverse until it reaches a Merge or Replicated Merge table. So if there is a Buffer table in the way, it will keep going.

Option 2: Distribute will forward the DDL to the table that it is immediately pointing to. If it is a Buffer table, so be it. now the DDL will be effectively executed on the Buffer table. The Buffer table will then forward this request to its sub table, be it a Merge or Replicated Merge table.

The two approaches can be generalized as: if a table gives a "mutation not allowed" error, don't throw that error if a setting is enabled. Instead, forward the DDL to the sub table.

As a side note, treating Update and Delete as DDL is not semantically accurate. I understand that the syntax is intentional to call out the cost of a vertical store, but there is a difference between a record change and a schema change. Not having that distinction could cause confusion and complication.

**Comment by SaltTan** [comment]
And how do alter a Distributed table itself?

**Comment by schien-dong** [comment]
Right, this touches a point I made earlier, which is the DDLs for changing schema and for Updating and Deleting are syntactically similar even though they are semantically different. Although it's hard to tell by looking at "alter table distributed_table ...", the "mutation not allowed" error is thrown at some point - that's when the intercept should take place. If a schema change on the distributed table can be carried out without invoking that error, then we know to not forward the query. Hope this strategy makes sense and causes minimal changes to the codebase.
